### PR TITLE
Auto Balance.

### PR DESCRIPTION
**Pull request by qiuyesuifeng**
/cc @siddontang Not have finished yet,  can this do what we want to do?

### CODE REVIEW

File: server/balancer.go [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+	"github.com/juju/errors"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/pingcap/kvproto/pkg/raftpb"
+)
+
+// Operator is the interface to do some operations.
+type Operator interface {
+	// Do does the operator, if finished then return true.
+	Do(req *pdpb.RegionHeartbeatRequest) (bool, *pdpb.RegionHeartbeatResponse)
```
Comment (siddontang) [inline server/balancer.go]: using `region* metapb.Region, leader* metapb.Peer` is better for argument.

using `*pdpb.ChangePeer` is better for return.

File: server/balancer.go [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+			},
+		}
+	*/
+
+	return false, nil
+}
+
+// Balancer is an interface to select store regions for auto-balance.
+type Balancer interface {
+	Balance(cluster *ClusterInfo, c *raftCluster) (*BalanceOperator, error)
```
Comment (siddontang) [inline server/balancer.go]: we can get cluster from raftCluster directly, so this interface argument seems redundant.

File: server/balancer.go [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+	return nil
+}
+
+// BalanceOperator is used to do region balance.
+type BalanceOperator struct {
+	index int
+	ops   []Operator
+}
+
+func newBalanceOperator(oldLeader, newLeader, newPeer *metapb.Peer) *BalanceOperator {
```
Comment (siddontang) [inline server/balancer.go]: This seems for capacity operator only.
As a common abstract, maybe you should create it with Operator directly, like 
```
func newBalanceOperator(ops ...Operator) *BalanceOperator
```

And you should combine the operators outer for special strategy.

File: server/balancer.go [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+	bo.index++
+	return false, nil
+}
+
+// ChangePeerOperator is used to do peer change.
+type ChangePeerOperator struct {
+	changePeer *pdpb.ChangePeer
+}
+
+func newChangePeerOperator(changePeer *pdpb.ChangePeer) *ChangePeerOperator {
```
Comment (siddontang) [inline server/balancer.go]: you can specialize `newAddPeerOperator` and `newRemovePeerOperator` for outer use easily. 
So outer doesn't need to construct `pdpb.ChangePeer` directly.

File: server/balancer.go [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+import (
+	"github.com/golang/protobuf/proto"
+	"github.com/juju/errors"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/pingcap/kvproto/pkg/raftpb"
+)
+
+// Operator is the interface to do some operations.
+type Operator interface {
```
Comment (siddontang) [inline server/balancer.go]: I think we should return the error to cancel this region balance too.

File: server/balancer.go [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+	if region == nil {
+		return false
+	}
+
+	leaderPeer := req.GetLeader()
+	if leaderPeer == nil {
+		return false
+	}
+
+	if co.changePeer.GetChangeType() == raftpb.ConfChangeType_AddNode {
```
Comment (siddontang) [inline server/balancer.go]: should we check epoch here? 
E,g, region has 1, 2, if we want to add 3, but the report is 1,2, 4, how to do it now?
I think we should do some check to cancel this balance if possible.

File: server/balancer.go [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+func newTransferLeaderOperator(oldLeader, newLeader *metapb.Peer) *TransferLeaderOperator {
+	return &TransferLeaderOperator{
+		oldLeader: oldLeader,
+		newLeader: newLeader,
+	}
+}
+
+// Check implements Operator.Check interface.
+func (lto *TransferLeaderOperator) Check(req *pdpb.RegionHeartbeatRequest) bool {
+	leader := req.GetLeader()
```
Comment (siddontang) [inline server/balancer.go]: how about leader is another? 
e.g, we want to do 1 -> 2, but now the leader is 3.

File: server/balancer.go [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+	storeID := store.store.GetId()
+	return peers[storeID]
+}
+
+func (cb *capacityBalancer) Balance(cluster *ClusterInfo, c *raftCluster) (*BalanceOperator, error) {
+	// Firstly, select one balance region from cluster info.
+	stores := cluster.getStores()
+	region, oldLeader := cb.SelectBalanceRegion(stores, cluster.regions)
+	if region == nil || oldLeader == nil {
+		return nil, errors.New("Region cannot be found to do balance")
```
Comment (siddontang) [inline server/balancer.go]: if we don't select one, why returning error here?

File: server/balancer.go [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+
+// Balancer is an interface to select store regions for auto-balance.
+type Balancer interface {
+	Balance(cluster *ClusterInfo, c *raftCluster) (*BalanceOperator, error)
+}
+
+const (
+	// If the used ratio of one storage is greater than this value,
+	// it will never be used as a selected target and it should be rebalanced.
+	maxCapacityUsedRatio = 0.4
```
Comment (siddontang) [inline server/balancer.go]: We should care more for capacity balance, not only UsedRatio.
E,g, all has 1T, 1,2,3 has 300G, but 4 has nothing, do we need to do balance?

File: server/balancer.go (hunk 7 above)
Comment (qiuyesuifeng) [inline server/balancer.go]: In balance workflow, we should select the older leader and new leader peer, so maybe it is needless to process this kind of case.
If you want 1 -> 2, then we should only send cmd to the store which leader 1 exists to do leader transfer.

File: server/balancer.go (hunk 2 above)
Comment (siddontang) [inline server/balancer.go]: you should return the balance region too. or you can embed this in BalanceOperator. 
Outer must know which region to balance.

File: server/balancer.go (hunk 6 above)
Comment (qiuyesuifeng) [inline server/balancer.go]: I have not checked epoch now, you are right, we should check the epoch to cancel the balance.

File: server/balancer.go (hunk 8 above)
Comment (qiuyesuifeng) [inline server/balancer.go]: An error returned or a nil value seems not different, you can ignore it outside.
You prefer to return nil value here?

File: server/balancer.go (hunk 9 above)
Comment (qiuyesuifeng) [inline server/balancer.go]: Agreed, i also notice this, process it in this PR or next, what is your opinion?

File: server/balancer.go (hunk 7 above)
Comment (siddontang) [inline server/balancer.go]: you select the old and new leader before, but when we execute the op, old leader may change from 1 -> 3, so the check will return false and following Do can not transfer 1 -> 2 too.

File: server/balancer.go (hunk 8 above)
Comment (siddontang) [inline server/balancer.go]: I prefer log a warning and return nil. 
this is may an error.

File: server/balancer.go (hunk 9 above)
Comment (siddontang) [inline server/balancer.go]: next PR, and then we should add unit test for these strategies.

File: server/balancer.go (hunk 2 above)
Comment (qiuyesuifeng) [inline server/balancer.go]: If we return a region value, seems we can check epoch version outside?
Then operator check will be more simple. Can this work?

File: server/balancer.go (hunk 7 above)
Comment (qiuyesuifeng) [inline server/balancer.go]: Yes, that is a question, if that, we should continue the balance step although the new leader is not what we want.

**Comment by siddontang** [comment]
PTAL @disksing @shenli

File: server/operator.go [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+	"github.com/pingcap/kvproto/pkg/raftpb"
+)
+
+// Operator is the interface to do some operations.
+type Operator interface {
+	// Do does the operator, if finished then return true.
+	Do(region *metapb.Region, leader *metapb.Peer) (bool, *pdpb.RegionHeartbeatResponse, error)
+	// Check checks whether the operator has been finished.
+	Check(region *metapb.Region, leader *metapb.Peer) (bool, error)
+}
```
Comment (disksing) [inline server/operator.go]: Is there a change that an Operator could never get complete?
For instance, balancer found a store to do 'AddPeer', then the store goes offline soon after.

Comment (qiuyesuifeng) [inline server/operator.go]: Now, if the store is offline, the Do method will return an error, outside worker loop will cancel the balance.
Maybe we can add error retry count for it later.

Comment (siddontang) [inline server/operator.go]: the raft can `AddPeer` successfully even the store doesn't exist...
Like @qiuyesuifeng said, we will support Cancel mechanism later. 
E.g, if the ops run not ok for a long time or many times, we can cancel it.

File: server/operator.go [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+func (lto *TransferLeaderOperator) Do(region *metapb.Region, leader *metapb.Peer) (bool, *pdpb.RegionHeartbeatResponse, error) {
+	ok, err := lto.Check(region, leader)
+	if err != nil {
+		return false, nil, errors.Trace(err)
+	}
+	if ok {
+		return true, nil, nil
+	}
+
+	// TODO: call admin command, maybe we can send it to a channel.
```
Comment (siddontang) [inline server/operator.go]: you can remove this comment now.

File: server/balancer.go [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+	return resultStore
+}
+
+func (cb *capacityBalancer) SelectBalanceRegion(stores []*StoreInfo, regions *RegionsInfo) (*metapb.Region, *metapb.Peer) {
+	store := cb.SelectFromStore(stores, true)
+	if store == nil {
+		return nil, nil
+	}
+
+	// Random select one leader region from store.
```
Comment (shenli) [inline server/balancer.go]: How about there is no leader peer on this store?

File: server/balancer.go [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+)
+
+type capacityBalancer struct {
+}
+
+func newCapacityBalancer() *capacityBalancer {
+	return &capacityBalancer{}
+}
+
+func (cb *capacityBalancer) SelectFromStore(stores []*StoreInfo, useFilter bool) *StoreInfo {
```
Comment (shenli) [inline server/balancer.go]: Why make it public?

File: server/balancer.go (hunk 9 above)
Comment (shenli) [inline server/balancer.go]: I suggest that we set two thresholds. The first one is soft threshold which is lower. The second one is hard threshold which is the same as  maxCapacityUsedRatio. If the used ratio is lower than soft threshold, we should ignore it.
The operator on region with used ratio lower than hard threshold should has low priority. How about we implement this strategy in the future PR?

File: server/balancer.go (hunk 13 above)
Comment (shenli) [inline server/balancer.go]: What's mean by useFilter?

Comment (shenli) [inline server/balancer.go]: I got it.

File: server/balancer.go [hunk 14]
```diff
@@ -0,0 +1,285 @@
...
+import (
+	"github.com/golang/protobuf/proto"
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	"github.com/pingcap/kvproto/pkg/metapb"
+)
+
+// Balancer is an interface to select store regions for auto-balance.
+type Balancer interface {
+	Balance(cluster *raftCluster) (*BalanceOperator, error)
```
Comment (shenli) [inline server/balancer.go]: How about we make raftCluster as member of balancer? Like https://play.golang.org/p/Joi2rYt9UC
So we could avoid passing it as argument between balancer inner functions.

Comment (siddontang) [inline server/balancer.go]: I prefer argument. The interface seems clear and we can know easily that we will use the clusterInfo to do the balance. :-)

@qiuyesuifeng

File: server/balancer.go [hunk 15]
```diff
@@ -0,0 +1,282 @@
...
+}
+
+func (cb *capacityBalancer) SelectToStore(stores []*StoreInfo, excluded map[uint64]struct{}) *StoreInfo {
+	var resultStore *StoreInfo
+	for _, store := range stores {
+		if store == nil {
+			continue
+		}
+
+		if _, ok := excluded[store.store.GetId()]; ok {
```
Comment (disksing) [inline server/balancer.go]: Any need to check `if excluded == nil`? L115 calls this func with a nil map.

Comment (qiuyesuifeng) [inline server/balancer.go]: Yes, it is my fault.

File: server/balancer.go (hunk 12 above)
Comment (disksing) [inline server/balancer.go]: @shenli It is checked at L164.

File: server/operator.go [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/pingcap/kvproto/pkg/raftpb"
+)
+
+var balancerIDs uint64
```
Comment (siddontang) [inline server/operator.go]: why need this?

File: server/balancer.go [hunk 17]
```diff
@@ -0,0 +1,438 @@
...
+var (
+	_ Balancer = &defaultBalancer{}
+	_ Balancer = &resourceBalancer{}
+)
+
+type resourceBalancer struct {
+	filters []Filter
+}
+
+func newResourceBalancer(minRatio float64, maxRatio float64) *resourceBalancer {
```
Comment (siddontang) [inline server/balancer.go]: I think later we can use a balancer config.

File: server/operator.go (hunk 16 above)
Comment (qiuyesuifeng) [inline server/operator.go]: It is only used to distinct different balancer.

File: server/balancer.go (hunk 17 above)
Comment (qiuyesuifeng) [inline server/balancer.go]: Agreed.

File: server/operator.go (hunk 16 above)
Comment (siddontang) [inline server/operator.go]: Using name is better if you want to distinguish different balancers.

Comment (qiuyesuifeng) [inline server/operator.go]: Seems not so many difference, if we use name, we have to build balancer name.

**Comment by siddontang** [comment]
LGTM

PTAL @disksing @shenli

File: server/balancer.go (hunk 17 above)
Comment (shenli) [inline server/balancer.go]: +1   Config file is a must.

File: server/filter.go [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+
+func newSnapCountFilter(maxSnapSendingCount uint32, maxSnapReceivingCount uint32) *snapCountFilter {
+	return &snapCountFilter{
+		maxSnapSendingCount:   maxSnapSendingCount,
+		maxSnapReceivingCount: maxSnapReceivingCount,
+	}
+}
+
+func (sf *snapCountFilter) FilterFromStore(store *storeInfo, args ...interface{}) bool {
+	return store.stats.stats.GetSnapSendingCount() > maxSnapSendingCount
```
Comment (disksing) [inline server/filter.go]: Do you mean `sf.maxSnapSendingCount` instead of `maxSnapSendingCount`?

Comment (qiuyesuifeng) [inline server/filter.go]: Yes, you are right.

Comment (disksing) [inline server/filter.go]: Maybe better rename the global consts to `defaultXxx`.

Comment (qiuyesuifeng) [inline server/filter.go]: Later we should use a config file for this arguments.
But i think `defaultMaxXXX` is a little strange, maybe `maxXXX` is better.

File: server/filter.go [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+// Filter is an interface to filter target store.
+type Filter interface {
+	FilterFromStore(store *storeInfo, args ...interface{}) bool
+	FilterToStore(store *storeInfo, args ...interface{}) bool
+}
```
Comment (disksing) [inline server/filter.go]: Please add some notes for the returned bool.

File: server/balancer.go [hunk 20]
```diff
@@ -0,0 +1,438 @@
...
+
+	if oldStoreScore <= newStoreScore {
+		log.Debugf("check score failed - old peer: %v, new peer: %v, old store score: %d, new store score :%d", oldPeer, newPeer, oldStoreScore, newStoreScore)
+		return false
+	}
+
+	// If the diff score is in defaultScoreFraction range, then we will do nothing.
+	diffScore := oldStoreScore - newStoreScore
+	if diffScore <= int(float64(oldStoreScore)*defaultDiffScoreFraction) {
+		log.Debugf("check score failed - diff score is too small - old peer: %v, new peer: %v, old store score: %d, new store score :%d, diif score: %d",
```
Comment (disksing) [inline server/balancer.go]: **diff** store

File: server/balancer.go [hunk 21]
```diff
@@ -0,0 +1,438 @@
...
+		return false
+	}
+
+	oldStoreScore := rb.score(oldStore, regionCount)
+	newStoreScore := rb.score(newStore, regionCount)
+
+	if oldStoreScore <= newStoreScore {
+		log.Debugf("check score failed - old peer: %v, new peer: %v, old store score: %d, new store score :%d", oldPeer, newPeer, oldStoreScore, newStoreScore)
+		return false
+	}
```
Comment (disksing) [inline server/balancer.go]: This check may be removed. It's covered by the diff check below.

File: server/balancer.go [hunk 22]
```diff
@@ -0,0 +1,438 @@
...
+
+	if isLeaderBalance {
+		ops, ok, err := rb.doLeaderBalance(cluster, stores, region, oldLeader)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if !ok {
+			// If we cannot find a region peer to do leader transfer,
+			// then we do follower balance instead.
+			return rb.doFollowerBalance(cluster, stores, region, oldLeader)
```
Comment (disksing) [inline server/balancer.go]: Why should we balance follower peer? The follower must not be on the Store we selected before.

File: server/balancer.go [hunk 23]
```diff
@@ -0,0 +1,438 @@
...
+			resultStore = store
+		}
+	}
+
+	return resultStore
+}
+
+// selectBalanceRegion try to select a store leader region to do balance and returns true, but if we cannot find any,
+// we try to find a store follower region and returns false.
+func (rb *resourceBalancer) selectBalanceRegion(cluster *clusterInfo, stores []*storeInfo) (*metapb.Region, *metapb.Peer, bool) {
```
Comment (disksing) [inline server/balancer.go]: ... **tries** to select...

File: server/balancer.go (hunk 21 above)
Comment (qiuyesuifeng) [inline server/balancer.go]: Yep.

**Comment by disksing** [comment]
LGTM

**Pull request merged**